### ISSUE TITLE
Create per call recordings that are not yet finished using a temporary extension

### ISSUE THREAD

**Issue by Guilhermerrrr** [lwvmobile/dsd-fme#172]
I'm running into a small problem with recordings from DSD FME that are being ingested by [Rdio Scanner](https://github.com/chuot/rdio-scanner). Rdio wants to ingest files that were created but not yet finished, which in turn streams only part of the transmission, once the call is over and the file is finally complete it tries to import it again but its prevented due to the duplicate file detection.

It would be nice if DSD FME could create a temporary file to write to like "file_name.wav.temp" while the call is still active, this way the streaming program can ignore this file if it's not ready yet. 

I should mention that Rdio Scanner does have a delay option that only imports files after a certain amount of seconds after the file is created but this still doesn't solve the problem, if this delay is too short it imports the file too soon, but if it is set to a very long time it creates an unnecessary long delay from the original transmission to the streaming moment.

I'd also want to take this opportunity to thank you @lwvmobile for the awesome and competent work you have done so far. All the hard work and the countless hours donated by you and everyone that contributes to this project are truly applaudable, and the impact of your efforts is invaluable. Thank you!

**Comment by lwvmobile** [comment]
Okay, thanks for the report. I'll look into it and see what can be done, but I don't have a time table for when I'll be able to work out a satisfactory fix for it, other than that it'll be on the 'TODO' list.

**Comment by lwvmobile** [comment]
Just as a followup question, what kind of system are you listening to with DSD-FME? It'll give me an area to start on when working on this issue, that way I can just focus on working per call for that system type and seeing what works best without having to work on all system types all at once.

**Comment by Guilhermerrrr** [comment]
> I don't have a time table for when I'll be able to work out a satisfactory fix for it

Thank you, no need to rush it, I just thought it would be a nice improvement since the two applications (DSD-FME and Rdio Scanner) could work really well in tandem.

> what kind of system are you listening to

It's a conventional DMR TII repeater 
I'm using this command to run DSD-FME v20230915 + aw240-20231104 patch
`dsd-fme-aw-aero.exe -i tcp -fs -Z -b XXX -N 2> log.ans -G group.csv`

Feel free to ask anything else, I'm happy to help.

**Comment by lwvmobile** [comment]
I was reviewing some of the code earlier to see what all changes could be made, and in order to permanently fix things, it would require a substantial rewrite of the call history and per call wav file creation, since the two things are tied together (for better or worse). If you only need this working together with a single DMR TII system, I can probably make a few changes here and there and have it opening a temporary file without an extension, and then when closing it, close it with the .wav extension, hopefully, Rdio Scanner doesn't ingest files without extensions, even if they are basically still wav files in all but extension.

Currently, because I didn't foresee this sort of issue, wav files are opened when the radio id changes, and are left open indefinitely until the radio id changes again, at which time that file is closed and a new wav file is opened, so I imagine that is a good part of the issue. I can rewrite things slightly so that the files are opened when the VLC arrives, and close them when a TLC arrives. Most DMR TII systems have those, and if you have good reception it shouldn't be an issue, but they may not exist on some TIII systems, especially on private calls.

In short, I think I can turn out a patched exe for you that should work like this sooner than later, and would only work on DMR TII while breaking things on any other system type (NXDN, P25, etc), but a more permanent solution I honestly have no idea how long that would take, and I am not very eagerly looking forward to re-writing that large a chunk of code with  holidays coming up.

**Comment by Guilhermerrrr** [comment]
Wow I really appreciate your effort, I want to emphasize that there's absolutely no rush on my end. Take all the time you need to work on the project, I understand that these things require time and effort. But sure, I'd be more than willing to test a modified exe and give you feedback. 
>would only work on DMR TII while breaking things on any other system type

There are no other system types in my area so that's not a problem

>when the VLC arrives, and close them when a TLC arrives. Most DMR TII systems have those 

I can provide you with baseband recordings if it helps.

Thank you, again

**Comment by lwvmobile** [comment]
> There are no other system types in my area so that's not a problem

Okay, well, that'll make things easier/quicker for a temporary solution.



> I can provide you with baseband recordings if it helps.

I have a feeling that may not be necessary, but if you did want to submit any samples, baseband recordings would be severely overkill, especially for a single system. You can use the built in symbol capture bin file format from DSD-FME if you want to share a sample. What you can do is start up with the `-c capture.bin` argument, or while focused on the ncurses terminal you can hit the `R` key to start a capture bin, and then hit the `r` key later on to stop it, and it'll place a capture.bin file in your dsd-fme folder for playback later on. They are relatively small in size, and if you zip them, they compress quite well.

**Comment by Guilhermerrrr** [comment]
Ok, I'll try to generate them tomorrow. I also took the liberty to DM you on Radio Reference. Thank you and good night

**Comment by Guilhermerrrr** [comment]
Unfortunately I just saw a post from Rdio Scanner owner from a few days ago saying he will ditch the dirwatch feature and the program will be solely based on API https://github.com/chuot/rdio-scanner/issues/393#issuecomment-1808397836 I appreciate the effort but I don't think you should waste time changing the code just for my use case, it's better to improve the whole thing when you get the chance

**Comment by lwvmobile** [comment]
Just out of curiosity, and a bit of user end testing, what happens if you take a wav file, remove its extension, and copy and paste it into the dirwatch directory, does it attempt to injest it, or does the file remain there. If the file remains there indefinitely, and then if you were to rename it again with the .wav extension, does it get injested at that point?

**Comment by Guilhermerrrr** [comment]
![Screenshot_20231119-002243_Firefox](https://github.com/lwvmobile/dsd-fme/assets/92966143/a00c1275-1a7c-477b-9915-cb147e276e9a)
in this screenshot you can see the user has to discriminate which extension to import, presumably it will ignore everything else. And if the file were renamed it would be imported normally

**Comment by lwvmobile** [comment]
Okay, well that's good that it can filter by extension and doesn't attempt to autodetect filetypes without extension (that's pretty common in Linux anyways). Anyways, if you get a chance, give this patched exe file a try, and see if it'll work for you. I dabbled with it a bit yesterday and today, and have it opening wav files on the VLC, or embedded voice link control, and then closing them on the TLC with an extension added. Keep in mind, it'll only work on DMR currently, and breaks functionality in other system types. Think of it more as a short term fix than anything else. 

[dsd-fme-aw252g-20231119-patch.zip](https://github.com/lwvmobile/dsd-fme/files/13402626/dsd-fme-aw252g-20231119-patch.zip)

**Comment by Guilhermerrrr** [comment]
Ok I'll try it and report back. Just to be sure, I have to copy this exe to `dsd-fme-20230915-win32\dsd-fme` folder and change the .bat to point to dsd-fme-aw-g.exe file, right? And using the Ncurses terminal go to `Input & output Options > Save per call decoded WAV`

**Comment by lwvmobile** [comment]
Yeah, that'll be fine, or you can use the -P switch to save yourself a step each time.
` dsd-fme-aw-g.exe -i tcp -fs -Z -b XXX -P -N 2> log.ans -G group.csv` to save yourself a step.

**Comment by Guilhermerrrr** [comment]
It works! Thank you!

**Comment by lwvmobile** [comment]
Alright, that's good to know. Let me know if any little issues occur though, like random crashes if it tries to write to a phantom wav file or something.  I'm not considering that exe file to be a definitive patch or anything, its just good enough (hopefully) to get that issue working for you on your single DMR TII frequency.

**Comment by lwvmobile** [comment]
I'm going to go ahead and close this issue as completed for now, keeping in mind a more comprehensive overhaul of the code that governs per call will be worked on sometime in the future, its just not going to be anytime soon.

**Comment by Dmedia3412** [comment]
Is it possible to do the same but for a regular conventional DMR?
I am also experiencing this problem with Rdioscanner.

**Comment by lwvmobile** [comment]
> Is it possible to do the same but for a regular conventional DMR?
> I am also experiencing this problem with Rdioscanner.

Try to use the -aw-guilhermerrrr version in the latest release zip file and see if that works for you. It should work on any DMR as long as there are VLC and TLC frames at the beginning and end of each TX. Tier 3 is the only place where that may not occur, depending on vendor.

**Comment by Dmedia3412** [comment]
Files do not become WAV at the end of the call.

This is how I run. And even if I'm an administrator

dsd-fme-aw-guilhermerrrr.exe -i tcp:127.0.0.1:7355 -o null -fs -u5 -P -7 DSD1 -K examples\multi_key_hex.csv -N 2> voice1.log

**Comment by lwvmobile** [comment]
It may not work correctly with the -7 option, I honestly don't remember, but if it never changes the extension to the correct name later on, there must be no TLC on the end of the Simplex transmission, or something is going on.

**Comment by Dmedia3412** [comment]
Yes, it really doesn’t work with -7.

I'm running 4 instances of FME and would like to write to different folders.

**Comment by lwvmobile** [comment]
> Yes, it really doesn’t work with -7.
> 
> I'm running 4 instances of FME and would like to write to different folders.

I'll try to look at it sometime when I have a chance, but its not going to be very high priority or anything. These 'patches' aren't official, and are just to fix a very specific thing for a very specific use case. As already noted here and in other per call wav file threads, the whole call history/per call and etc needs to be reworked, and that is going to be a major task. Keep in mind, this was written to fix the rdio ingesting the wav files too early for a very specific circumstance.

**Comment by Dmedia3412** [comment]
I get it, I just need to create 4 different folders with FME. And in each folder use the WAV folder.
  I'll try this.

PS. Great, everything is as it should be now!

**Comment by lwvmobile** [comment]
Okay, well, I went ahead and fixed the code so that it'll behave with custom wav file directories now, in case you still want to use that feature without needing 4 copies of the entire folder set up, or just want to test it out. Keep in mind, this is still only for DMR when they use VLC or Embedded LC, and then TLC to close the wav files. It is not set up for use on P25, NXDN, etc, and if you attempt to use wav file output on those, the functionality will break, and may even crash.

[dsd-fme-aw-278g.zip](https://github.com/lwvmobile/dsd-fme/files/13678957/dsd-fme-aw-278g.zip)

This will probably be part of the larger scheme on per call and history, etc, just don't have a lot of time to sit down and spend re-writing and testing  a bunch of new code to govern all those things right now, so its quick and dirty for DMR Simplex and Tier 2, Tier 3 may also work, but needs VLC and TLC, some T3 systems use Illegally Parked and forego setting up the VLC, and as such, cause issues, or may skip the first moments of audio. Hytera XPT may also be a problem with its TLC voice call setup, not sure, not going to test that out either.

Also, to answer your other question, just use -fs for DMR only. This patched version also only writes the file as the CC variant, and not the MS variant for simplex, so that should also fix that issue for you.

**Comment by Dmedia3412** [comment]
Yes it works now.
The files become WAV only at the end of the transfer and it works -7
Thanks for your work!

This is the most convenient combination [DSD-FME]+[Rdio-Scanner].
